blocks/datasets/streams.py

```diff
@@ -314,6 +314,38 @@ def _cache(self):
             cache.extend(data)
 
 
+class SortMapping(object):
+    """Callable class for creating sort mappings without nesting functions."""
+    def __init__(self, key, reverse):
+        self.key = key
+        self.reverse = reverse
+
+    def __call__(self, x):
+        indices = [i for (v, i) in
+                   sorted(((v, i) for (i, v) in enumerate(x[0])),
+                          key=self.key, reverse=self.reverse)]
+        return tuple([[i[j] for j in indices] for i in x])
+
+
+class DataStreamSort(DataStreamMapping):
+    """Sorts the contents of the batches of the wrapped data stream.
+
+    Parameters
+    ----------
+    data_stream : instance of :class:`DataStream`
+        The wrapped data stream.
+    key : callable
+        The mapping that returns the value to sort on
+
+    """
+    def __init__(self, data_stream, key=None, reverse=False):
+
+        super(DataStreamSort, self).__init__(data_stream,
+                                             mapping=SortMapping(key, reverse))
+        self.key = key
```

Comment:
A user can always access it through e.g. `stream.mapping.key` right? Because `stream.mapping` is a class instance in this case, so I'd just get rid of it, but your call.